diff --git a/permissions_by_term.module b/permissions_by_term.module
index 6597a7a..f8fe772 100644
--- a/permissions_by_term.module
+++ b/permissions_by_term.module
@@ -290,9 +290,12 @@ function permissions_by_term_user_form_submit(array &$form, FormStateInterface $
 
   /** @var \Drupal\user\UserInterface $user */
   $user = $form_state->getFormObject()->getEntity();
+  $uid = $user->id();
+
+  $origPerms = $access_storage->getAllTermPermissionsByUserId($uid);
 
   // First, we delete existing values from the db.
-  $access_storage->deleteAllTermPermissionsByUserId($user->id());
+  $access_storage->deleteAllTermPermissionsByUserId($uid);
 
   if ($tids = $form_state->getValue('terms')) {
     // For term permissions use user preferred language.
@@ -300,13 +303,16 @@ function permissions_by_term_user_form_submit(array &$form, FormStateInterface $
 
     // Second, we insert updated values.
     foreach ($tids as $tid) {
-      $access_storage->addTermPermissionsByUserIds([$user->id()], $tid, $langcode);
+      $access_storage->addTermPermissionsByUserIds([$uid], $tid, $langcode);
     }
   }
 
+  $newPerms = $access_storage->getAllTermPermissionsByUserId($uid);
+
   // Rebuild permissions for nodes if needed.
-  if (!\Drupal::config('permissions_by_term.settings')->get('disable_node_access_records')) {
-    node_access_rebuild(TRUE);
+  if (!\Drupal::config('permissions_by_term.settings')->get('disable_node_access_records') &&
+    !empty(array_diff($origPerms, $newPerms))) {
+    \Drupal::service('permissions_by_term.node_access')->rebuildAccess($uid);
   }
 }
 
-- 
2.31.1.windows.1

